// react-md/lib/Helpers/HorizontalAnchors.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.HorizontalAnchors = mod.exports;
  }
})(this, function (exports) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  /** @module Helpers/HorizontalAnchors */

  /**
   * An enum for all the different types of horizontal anchors on a layover.
   *
   * @readonly
   * @enum {string}
   */
  var HorizontalAnchors = {
    /**
     * Positions the layover to the outside left on the toggle component
     * so that it does not overlap.
     */
    LEFT: 'left',

    /**
     * Positions the layover to the inner left of the toggle component so
     * that it will overlap.
     */
    INNER_LEFT: 'inner left',

    /**
     * Positions the layover so that it overlaps the toggle component
     * by positioning itself in the cetner of the toggle component's width.
     */
    CENTER: 'center',

    /**
     * Positions the layover to the outside right on the toggle component
     * so that it does not overlap.
     */
    RIGHT: 'right',

    /**
     * Positions the layover to the outside right on the toggle component
     * so that it will overlap.
     */
    INNER_RIGHT: 'inner right'
  };

  exports.default = HorizontalAnchors;
});
