// react-md/lib/Menus/Menu.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'prop-types', 'react-dom', 'classnames', 'react-prop-types/lib/deprecated', 'react-prop-types/lib/isRequiredForA11y', '../constants/CSSTransitionGroupTick', '../utils/getField', '../utils/EventUtils/handleKeyboardAccessibility', '../Helpers/anchorShape', '../Helpers/fixedToShape', '../Helpers/positionShape', '../Helpers/Layover', '../Lists/List'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('prop-types'), require('react-dom'), require('classnames'), require('react-prop-types/lib/deprecated'), require('react-prop-types/lib/isRequiredForA11y'), require('../constants/CSSTransitionGroupTick'), require('../utils/getField'), require('../utils/EventUtils/handleKeyboardAccessibility'), require('../Helpers/anchorShape'), require('../Helpers/fixedToShape'), require('../Helpers/positionShape'), require('../Helpers/Layover'), require('../Lists/List'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.reactDom, global.classnames, global.deprecated, global.isRequiredForA11y, global.CSSTransitionGroupTick, global.getField, global.handleKeyboardAccessibility, global.anchorShape, global.fixedToShape, global.positionShape, global.Layover, global.List);
    global.Menu = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _reactDom, _classnames, _deprecated, _isRequiredForA11y, _CSSTransitionGroupTick, _getField, _handleKeyboardAccessibility, _anchorShape, _fixedToShape, _positionShape, _Layover, _List) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _classnames2 = _interopRequireDefault(_classnames);

  var _deprecated2 = _interopRequireDefault(_deprecated);

  var _isRequiredForA11y2 = _interopRequireDefault(_isRequiredForA11y);

  var _CSSTransitionGroupTick2 = _interopRequireDefault(_CSSTransitionGroupTick);

  var _getField2 = _interopRequireDefault(_getField);

  var _handleKeyboardAccessibility2 = _interopRequireDefault(_handleKeyboardAccessibility);

  var _anchorShape2 = _interopRequireDefault(_anchorShape);

  var _fixedToShape2 = _interopRequireDefault(_fixedToShape);

  var _positionShape2 = _interopRequireDefault(_positionShape);

  var _Layover2 = _interopRequireDefault(_Layover);

  var _List2 = _interopRequireDefault(_List);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  function _objectWithoutProperties(obj, keys) {
    var target = {};

    for (var i in obj) {
      if (keys.indexOf(i) >= 0) continue;
      if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
      target[i] = obj[i];
    }

    return target;
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var Menu = function (_PureComponent) {
    _inherits(Menu, _PureComponent);

    function Menu() {
      var _ref;

      var _temp, _this, _ret;

      _classCallCheck(this, Menu);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Menu.__proto__ || Object.getPrototypeOf(Menu)).call.apply(_ref, [this].concat(args))), _this), _this._handleClose = function (e) {
        var _this$props = _this.props,
            close = _this$props.close,
            onClose = _this$props.onClose;

        if (close || onClose) {
          (close || onClose)(e);
        }
      }, _this._handleClick = function (e) {
        if (_this.props.onClick) {
          _this.props.onClick(e);
        }

        var node = e.target;
        while (_this._container && _this._container.contains(node)) {
          if (node.classList.contains('md-list-control')) {
            return;
          } else if (!node.classList.contains('md-list-item--nested-container') && node.classList.contains('md-list-item')) {
            e.persist();
            _this._timeout = setTimeout(function () {
              _this._timeout = null;
              _this._handleClose(e);
            }, _CSSTransitionGroupTick2.default);

            return;
          }

          node = node.parentNode;
        }
      }, _this._handleKeyDown = function (e) {
        if (_this.props.onKeyDown) {
          _this.props.onKeyDown(e);
        }

        (0, _handleKeyboardAccessibility2.default)(e, _this._handleClick, true, true);
      }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Menu, [{
      key: 'getChildContext',
      value: function getChildContext() {
        var _props = this.props,
            cascading = _props.cascading,
            id = _props.id,
            fixedTo = _props.fixedTo,
            cascadingAnchor = _props.cascadingAnchor;

        var listLevel = this.context.listLevel || 0;
        var cascadingMenu = typeof cascading !== 'undefined' ? cascading : this.context.cascadingMenu;
        var cascadingZDepth = (0, _getField2.default)(this.context, this.props, 'cascadingZDepth');
        var cascadingFixedTo = typeof fixedTo !== 'undefined' ? fixedTo : this.context.cascadingFixedTo;

        return {
          listLevel: listLevel,
          cascadingId: id + '-level-' + (listLevel + 1),
          cascadingMenu: cascadingMenu,
          cascadingAnchor: cascadingAnchor,
          cascadingZDepth: cascadingZDepth,
          cascadingFixedTo: cascadingFixedTo
        };
      }
    }, {
      key: 'componentDidMount',
      value: function componentDidMount() {
        this._container = (0, _reactDom.findDOMNode)(this);
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        if (this._timeout) {
          clearTimeout(this._timeout);
        }
      }
    }, {
      key: 'render',
      value: function render() {
        var _props2 = this.props,
            id = _props2.id,
            className = _props2.className,
            listStyle = _props2.listStyle,
            listClassName = _props2.listClassName,
            visible = _props2.visible,
            children = _props2.children,
            listProps = _props2.listProps,
            listZDepth = _props2.listZDepth,
            listInline = _props2.listInline,
            listHeightRestricted = _props2.listHeightRestricted,
            sameWidth = _props2.sameWidth,
            contained = _props2.contained,
            isOpen = _props2.isOpen,
            propFixedTo = _props2.fixedTo,
            propListId = _props2.listId,
            cascading = _props2.cascading,
            cascadingAnchor = _props2.cascadingAnchor,
            cascadingZDepth = _props2.cascadingZDepth,
            onClose = _props2.onClose,
            propPosition = _props2.position,
            close = _props2.close,
            autoclose = _props2.autoclose,
            limitHeight = _props2.limitHeight,
            expanderIconChildren = _props2.expanderIconChildren,
            expanderIconClassName = _props2.expanderIconClassName,
            props = _objectWithoutProperties(_props2, ['id', 'className', 'listStyle', 'listClassName', 'visible', 'children', 'listProps', 'listZDepth', 'listInline', 'listHeightRestricted', 'sameWidth', 'contained', 'isOpen', 'fixedTo', 'listId', 'cascading', 'cascadingAnchor', 'cascadingZDepth', 'onClose', 'position', 'close', 'autoclose', 'limitHeight', 'expanderIconChildren', 'expanderIconClassName']);

        var _props3 = this.props,
            listId = _props3.listId,
            position = _props3.position;

        if (!listId) {
          listId = id + '-list';
        }

        if (position === 'context') {
          position = Menu.Positions.BELOW;
        }

        var fixedTo = typeof propFixedTo !== 'undefined' ? propFixedTo : this.context.cascadingFixedTo;
        var listVisible = typeof isOpen !== 'undefined' ? isOpen : visible;
        return _react2.default.createElement(
          _Layover2.default,
          _extends({
            id: id,
            className: (0, _classnames2.default)('md-menu-container', className)
          }, props, {
            sameWidth: contained || sameWidth,
            fixedTo: fixedTo,
            onClick: this._handleClick,
            onKeyDown: this._handleKeyDown,
            onClose: this._handleClose,
            animationPosition: position,
            visible: listVisible,
            'aria-haspopup': true,
            'aria-expanded': listVisible,
            'aria-owns': listId
          }),
          _react2.default.createElement(
            _List2.default,
            _extends({}, listProps, {
              id: listId,
              key: 'menu-list',
              style: listStyle,
              className: (0, _classnames2.default)('md-list--menu', _defineProperty({
                'md-list--menu-restricted': listHeightRestricted
              }, 'md-paper md-paper--' + listZDepth, listZDepth), listClassName),
              inline: listInline
            }),
            children
          )
        );
      }
    }]);

    return Menu;
  }(_react.PureComponent);

  Menu.HorizontalAnchors = _Layover2.default.HorizontalAnchors;
  Menu.VerticalAnchors = _Layover2.default.VerticalAnchors;
  Menu.Positions = {
    // Can't do ...Layover.Positions since it triggers the get for CONTEXT
    TOP_LEFT: _Layover2.default.Positions.TOP_LEFT,
    TOP_RIGHT: _Layover2.default.Positions.TOP_RIGHT,
    BOTTOM_LEFT: _Layover2.default.Positions.BOTTOM_LEFT,
    BOTTOM_RIGHT: _Layover2.default.Positions.BOTTOM_RIGHT,
    BELOW: _Layover2.default.Positions.BELOW,
    _warned: false,
    get CONTEXT() {
      if (!this._warned) {
        /* eslint-disable no-console */
        console.error('The `Menu.Positions.CONTEXT` position has been depreacted and will be removed ' + 'in the next major release. To make the `Menu` behave as a context menu, provide ' + 'the `onContextMenu` prop instead.');
      }

      this._warned = true;
      return 'context';
    }
  };
  Menu.propTypes = {
    /**
     * An id to provide to the menu's container. This is required for accessibility as it generates
     * the `aria-` attributes for dynamic content.
     *
     * @see {@link #listId}
     */
    id: (0, _isRequiredForA11y2.default)(_propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string])),

    /**
     * An optional id to provide to the menu's list. If this prop is omitted, the list's id will be
     * `\`${id}-list\``
     */
    listId: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),

    /**
     * An optional style to apply to the main container for the menu.
     */
    style: _propTypes2.default.object,

    /**
     * An optional class name to apply to the main container for the menu.
     */
    className: _propTypes2.default.string,

    /**
     * An optional style to apply to the list once the menu has opened.
     */
    listStyle: _propTypes2.default.object,

    /**
     * An optional class name to apply to the list once the menu has opened.
     */
    listClassName: _propTypes2.default.string,

    /**
     * The component to render the main container as.
     *
     * @see {@link Helpers/Layovers#component}
     */
    component: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]),

    /**
     * This is how the menu's `List` gets anchored to the `toggle` element.
     *
     * @see {@link Helpers/Layovers#anchor}
     */
    anchor: _anchorShape2.default,

    /**
     * This is the optional anchor to use when the `position` is set to `Menu.Positions.BELOW`.
     * Set this to `null` to use the default `anchor` prop.
     *
     * @see {@link Helpers/Layovers#belowAnchor}
     */
    belowAnchor: _anchorShape2.default,

    /**
     * This is the animation position for the list that appears.
     *
     * @see {@link Helpers/Layovers#animationPosition}
     */
    position: _positionShape2.default,

    /**
     * This is the component/element that should toggle the menu open.
     *
     * @see {@link Helpers/Layovers#toggle}
     */
    toggle: _propTypes2.default.node,

    /**
     * This is how the menu's list will be "fixed" to the `toggle` component.
     *
     * @see {@link Helpers/Layovers#fixedTo}
     */
    fixedTo: _fixedToShape2.default,

    /**
     * Any additional props that should be applied to the list in the menu. This is really used
     * when additional `aria-` tags need to be applied.
     */
    listProps: _propTypes2.default.object,

    /**
     * Boolean if the menu's list should appear horizontally instead of vertically.
     */
    listInline: _propTypes2.default.bool,

    /**
     * The list's z-depth for applying box shadow. This should be a number from 0 to 5.
     */
    listZDepth: _propTypes2.default.number.isRequired,

    /**
     * Boolean if the list should have its height restricted to the `$md-menu-mobile-max-height`/
     * `$md-menu-desktop-max-height` values.
     *
     * @see [md-menu-mobile-max-height](/components/menus?tab=2#variable-md-menu-mobile-max-height)
     * @see [md-menu-desktop-max-height](/components/menus?tab=2#variable-md-menu-desktop-max-height)
     */
    listHeightRestricted: _propTypes2.default.bool,

    /**
     * Boolean if the menu's list is visible.
     */
    visible: _propTypes2.default.bool.isRequired,

    /**
     * Any children to render in the menu's list. This _should_ normally be `ListItem`, or
     * `ListItemControl`.
     */
    children: _propTypes2.default.node,

    /**
     * An optional function to call when en element in the menu has been clicked.
     */
    onClick: _propTypes2.default.func,

    /**
     * An optional function to call when a key is pressed anywhere in the menn.
     */
    onKeyDown: _propTypes2.default.func,

    /**
     * A function to call to close the menu. This is used for closing on outside clicks,
     * closing when a list item has been clicked, or the user presses escape.
     */
    onClose: _propTypes2.default.func.isRequired,

    /**
     * Boolean if the menu should be cascading. This means that the menu will pop the additional
     * `nestedItems` on any `ListItem` to be appear either to the right or left of the visible list.
     */
    cascading: _propTypes2.default.bool,

    /**
     * This is how the cascading lists get anchored to the list item.
     *
     * @see {@link Helpers/Layovers#anchor}
     */
    cascadingAnchor: _Layover2.default.propTypes.anchor,

    /**
     * This is the z-depth the list should gain for a cascading menu. This only gets applied on
     * items that are more than 1 level deep.
     */
    cascadingZDepth: _propTypes2.default.number.isRequired,

    /**
     * Boolean if the `md-full-width` class name should get applied to the menu's container.
     */
    fullWidth: _propTypes2.default.bool,

    /**
     * Boolean if the menu should be displayed as a block instead of as an inline block.
     *
     * @see {@link #fullWidth}
     */
    block: _propTypes2.default.bool,

    /**
     * @see {@link Helpers/Layovers#centered}
     */
    centered: _Layover2.default.propTypes.centered,

    /**
     * @see {@link Helpers/Layovers#sameWidth}
     */
    sameWidth: _Layover2.default.propTypes.sameWidth,

    /**
     * If you would like the menu to interact as a context menu, provide this prop.
     *
     * @see {@link Helpers/Layovers#onContextMenu}
     */
    onContextMenu: _Layover2.default.propTypes.onContextMenu,

    /**
     * Boolean if the default behavior of the context menu should be prevented when using the
     * `onContextMenu` prop.
     *
     * @see {@link Helpers/Layovers#preventContextMenu}
     */
    preventContextMenu: _Layover2.default.propTypes.preventContextMenu,

    /**
     * @see {@link Helpers/Layovers#xThreshold}
     */
    xThreshold: _propTypes2.default.number,

    /**
     * @see {@link Helpers/Layovers#yThreshold}
     */
    yThreshold: _propTypes2.default.number,

    /**
     * @see {@link Helpers/Layovers#closeOnOutsideClick}
     */
    closeOnOutsideClick: _propTypes2.default.bool,

    /**
     * @see {@link Helpers/Layovers#toggleQuery}
     */
    toggleQuery: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.object, _propTypes2.default.string]),

    /**
     * An optional transition name to use for the list appearing/disappearing.
     *
     * @see {@link Helpers/Layovers#transitionName}
     */
    transitionName: _propTypes2.default.string,

    /**
     * @see {@link Helpers/Layovers#transitionEnterTimeout}
     */
    transitionEnterTimeout: _propTypes2.default.number,

    /**
     * @see {@link Helpers/Layovers#transitionLeaveTimeout}
     */
    transitionLeaveTimeout: _propTypes2.default.number,

    /**
     * Boolean if the menu should automatically try to reposition itself to stay within
     * the viewport when the `fixedTo` element scrolls.
     *
     * @see {@link Helpers/Layovers#fixedTo}
     */
    repositionOnScroll: _propTypes2.default.bool,

    isOpen: (0, _deprecated2.default)(_propTypes2.default.bool, 'Use `visible` instead'),
    close: (0, _deprecated2.default)(_propTypes2.default.func, 'Use `onClose` instead'),
    autoclose: (0, _deprecated2.default)(_propTypes2.default.bool, 'The menus will always autoclose as according to the specs'),
    contained: (0, _deprecated2.default)(_propTypes2.default.bool, 'Use `sameWidth` instead'),
    limitHeight: (0, _deprecated2.default)(_propTypes2.default.bool, 'The menus will always be limited in height as according to the specs'),
    expanderIconClassName: (0, _deprecated2.default)(_propTypes2.default.node, 'The expander for cascading menus will now just be a simple rotate of the existing `ListItem` ' + 'expander icon'),
    expanderIconChildren: (0, _deprecated2.default)(_propTypes2.default.node, 'The expander for cascading menus will now just be a simple rotate of the existing `ListItem` ' + 'expander icon')
  };
  Menu.defaultProps = {
    anchor: {
      x: _Layover2.default.HorizontalAnchors.INNER_RIGHT,
      y: _Layover2.default.VerticalAnchors.OVERLAP
    },
    cascadingAnchor: {
      x: _Layover2.default.HorizontalAnchors.RIGHT,
      y: _Layover2.default.VerticalAnchors.OVERLAP
    },
    position: _Layover2.default.Positions.TOP_RIGHT,
    fixedTo: typeof window !== 'undefined' ? window : {},
    listZDepth: 2,
    listHeightRestricted: true,
    cascadingZDepth: 3,
    repositionOnScroll: true
  };
  Menu.contextTypes = {
    listLevel: _propTypes2.default.number,
    cascadingId: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
    cascadingMenu: _propTypes2.default.bool,
    cascadingAnchor: _anchorShape2.default,
    cascadingZDepth: _propTypes2.default.number
  };
  Menu.childContextTypes = {
    listLevel: _propTypes2.default.number,
    cascadingId: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
    cascadingMenu: _propTypes2.default.bool,
    cascadingFixedTo: _fixedToShape2.default,
    cascadingAnchor: _anchorShape2.default,
    cascadingZDepth: _propTypes2.default.number
  };
  exports.default = Menu;
});
