// react-md/lib/DataTables/contextTypes.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'prop-types'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('prop-types'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.propTypes);
    global.contextTypes = mod.exports;
  }
})(this, function (exports, _propTypes) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _propTypes2 = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports.default = {
    checkedIconChildren: _propTypes2.default.node,
    checkedIconClassName: _propTypes2.default.string,
    uncheckedIconChildren: _propTypes2.default.node,
    uncheckedIconClassName: _propTypes2.default.string,
    indeterminateIconChildren: _propTypes2.default.node,
    indeterminateIconClassName: _propTypes2.default.string,
    indeterminate: _propTypes2.default.bool,
    plain: _propTypes2.default.bool,
    selectableRows: _propTypes2.default.bool.isRequired,
    allSelected: _propTypes2.default.bool.isRequired,
    selectedRows: _propTypes2.default.arrayOf(_propTypes2.default.bool).isRequired,
    createCheckbox: _propTypes2.default.func.isRequired,
    removeCheckbox: _propTypes2.default.func.isRequired,
    toggleSelectedRow: _propTypes2.default.func.isRequired,
    baseId: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
    baseName: _propTypes2.default.string,
    checkboxHeaderLabel: _propTypes2.default.string.isRequired,
    checkboxLabelTemplate: _propTypes2.default.string.isRequired,
    fixedHeader: _propTypes2.default.bool.isRequired,
    fixedFooter: _propTypes2.default.bool.isRequired
  };
});
