// react-md/lib/utils/DateUtils/DateTimeFormat.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports);
    global.DateTimeFormat = mod.exports;
  }
})(this, function (exports) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  /** @module utils/DateUtils/DateTimeFormat */

  /**
   * A _really_ terrible DateTimeFormat mock if it is not available in
   * the browser or node version.
   */
  var DateTimeFormat = function () {
    if (typeof Intl !== 'undefined' && typeof Intl.DateTimeFormat !== 'undefined') {
      return Intl.DateTimeFormat;
    }

    // Very bad 'mock' of Intl.DateTimeFormat
    return function () {
      // eslint-disable-line no-shadow
      function DateTimeFormat(locales, formatOptions) {
        _classCallCheck(this, DateTimeFormat);

        this.locales = locales;
        this.options = formatOptions;
      }

      _createClass(DateTimeFormat, [{
        key: 'format',
        value: function format(date) {
          return date.toLocaleString();
        }
      }]);

      return DateTimeFormat;
    }();
  }();

  exports.default = DateTimeFormat;
});
